packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js

```diff
@@ -0,0 +1,191 @@
+// @ts-check
+import { assert, details } from '@agoric/assert';
+import { makeTranscriptManager } from './transcript';
+import { createSyscall } from './syscall';
+
+// eslint-disable-next-line no-unused-vars
+function parentLog(first, ...args) {
+  // console.error(`--parent: ${first}`, ...args);
+}
+
+const trace = label => x => {
+  parentLog(label, x);
+  return x;
+};
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+/**
+ * @param {{
+ *   startXSnap: (name: string, handleCommand: SyncHandler) => { worker: XSnap, bundles: Record<string, ExportBundle> },
+ *   kernelKeeper: KernelKeeper,
+ *   testLog: (...args: unknown[]) => void,
+ *   decref: (vatID: unknown, vref: unknown, count: number) => void,
+ * }} tools
+ * @returns { VatManagerFactory }
+ *
+ * @typedef { { moduleFormat: 'getExport', source: string } } ExportBundle
+ * @typedef { (msg: Uint8Array) => Uint8Array } SyncHandler
+ * @typedef { ReturnType<typeof import('@agoric/xsnap').xsnap> } XSnap
+ * @typedef { ReturnType<typeof import('../state/kernelKeeper').default> } KernelKeeper
+ * @typedef { ReturnType<typeof import('./manager-nodeworker').makeNodeWorkerVatManagerFactory> } VatManagerFactory
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ */
+export function makeXsSubprocessFactory({
+  startXSnap,
+  kernelKeeper,
+  testLog,
+  decref,
+}) {
+  /**
+   * @param { unknown } vatID
+   * @param { unknown } bundle
+   * @param { ManagerOptions } managerOptions
+   */
+  async function createFromBundle(vatID, bundle, managerOptions) {
+    parentLog('createFromBundle', { vatID });
+    const { vatParameters, virtualObjectCacheSize } = managerOptions;
+    assert(!managerOptions.metered, 'not supported yet');
+    assert(!managerOptions.enableSetup, 'not supported at all');
+    if (managerOptions.enableInternalMetering) {
+      // TODO: warn+ignore, rather than throw, because the kernel enables it
+      // for all vats, because the Spawner still needs it. When the kernel
+      // stops doing that, turn this into a regular assert
+      console.log(`xsnap worker does not support enableInternalMetering`);
+    }
+    const vatKeeper = kernelKeeper.getVatKeeper(vatID);
+    const transcriptManager = makeTranscriptManager(
+      kernelKeeper,
+      vatKeeper,
+      vatID,
+    );
+
+    const { doSyscall, setVatSyscallHandler } = createSyscall(
+      transcriptManager,
+    );
+
+    /** @type { (vatSyscallObject: Tagged) => unknown } */
+    function handleSyscall(vatSyscallObject) {
+      return trace('doSyscall')(doSyscall(vatSyscallObject));
+    }
+
+    /** @type { (vref: unknown, count: number) => void } */
+    function vatDecref(vref, count) {
+      decref(vatID, vref, count);
+    }
+
+    /** @type { (item: Tagged) => unknown } */
+    function handleUpstream([type, ...args]) {
```

Comment:
FWIW, the patterm I'm using is vat-centric. When the vat creates a Remotable and sends it to the kernel, we call that "export", and since that's the ego-centric thing a vat can do, it claims the `o+NN` side of the c-list numberspace. When the kernel sends a reference into a vat, i.e. when the vat receives a reference from the kernel (and creates a Presence / proxy), we call that "import", and it gets a negative number in the c-list numberspace. "syscall" and "dispatch/delivery" are the two directions of the messages across the kernel-vat boundary, for which "syscall" has convenient precedents in unix.

For this kernel-worker boundary, though, I'd be happy with having the pattern being that the kernel is in control, and the worker only does what the kernel says (and, in particular, is entirely idle until the kernel gives it an instruction). So "command" and "response" sound like good nouns to me. But, while the command is running, the worker gets to emit messages and block for a response, so we do need two pairs of nouns. "request" might usefully imply that the thing doing the request (the worker) is less commanding than the thing providing the answer (the kernel).

I dunno. Words are hard.

Reply:
Consistency with `manager-subprocess-node.js` and `manager-subprocess-xsnap.js` seems worthwhile. I'm leaving this alone until we figure out a change for all 3.